File: libs/typeorm/src/entities/RequestedSong.ts
import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryColumn, PrimaryGeneratedColumn,
  Relation,
} from 'typeorm';

import { type Channel } from './Channel.js';
import { type User } from './User.js';

@Entity('channels_requested_songs')
export class RequestedSong {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('varchar')
  videoId: string;

  @Column('text')
  title: string;

  @Column('int4')
  duration: number;

  @CreateDateColumn()
  createdAt: Date;

  @Column()
  orderedById: string;

  @Column()
  orderedByName: string;

  @Column({ nullable: true })
  orderedByDisplayName: string | null;

  @Column()
  queuePosition: number;

  @ManyToOne('User', 'requestedSongs')
  @JoinColumn({ name: 'orderedById' })
  orderedBy?: Relation<User>;

  @Column()
  channelId: string;

  @ManyToOne('Channel', 'requestedSongs')
  @JoinColumn({ name: 'channelId' })
  channel?: Relation<Channel>;

  @DeleteDateColumn({ nullable: true })
  deletedAt: Date | null;
}

File: libs/gomodels/requested_song.go
package model

import (
	"database/sql"
	"time"

	"github.com/guregu/null"
	uuid "github.com/satori/go.uuid"
)

var (
	_ = time.Second
	_ = sql.LevelDefault
	_ = null.Bool{}
	_ = uuid.UUID{}
)

type RequestedSong struct {
	ID                   string      `gorm:"primary_key;AUTO_INCREMENT;column:id;type:TEXT;" json:"id"`
	ChannelID            string      `gorm:"column:channelId;type:TEXT;"                     json:"channelId"`
	OrderedById          string      `gorm:"column:orderedById;type:TEXT;"                   json:"orderedById"`
	OrderedByName        string      `gorm:"column:orderedByName;type:TEXT;"                 json:"orderedByName"`
	OrderedByDisplayName null.String `gorm:"column:orderedByDisplayName;type:TEXT;"                 json:"orderedByDisplayName"`
	VideoID              string      `gorm:"column:videoId;type:varchar;"                    json:"videoId"`
	Title                string      `gorm:"column:title;type:text;"                         json:"title"`
	Duration             int32       `gorm:"column:duration;type:int4"                       json:"duration"`
	CreatedAt            time.Time   `gorm:"column:createdAt;type:TIMESTAMP;"                json:"createdAt"`
	DeletedAt            *time.Time  `gorm:"column:deletedAt;type:TIMESTAMP;"                json:"deletedAt"`
	QueuePosition        int         `gorm:"column:queuePosition;type:int4"                  json:"queuePosition"`

	// internal variable, not db
	Count int `gorm:"column:count" db:"count"`
}

func (c *RequestedSong) TableName() string {
	return "channels_requested_songs"
}
